domain/agentprovisioner/service/service.go

```diff
@@ -0,0 +1,260 @@
+		config.AuthorizedKeysKey,
+		config.EnableOSRefreshUpdateKey,
+		config.EnableOSUpgradeKey,
+		config.TypeKey,
+		config.SSLHostnameVerificationKey,
+		config.HTTPProxyKey, config.HTTPSProxyKey, config.FTPProxyKey, config.NoProxyKey,
+		config.JujuHTTPProxyKey, config.JujuHTTPSProxyKey, config.JujuFTPProxyKey, config.JujuNoProxyKey,
+		config.AptHTTPProxyKey, config.AptHTTPSProxyKey, config.AptFTPProxyKey, config.AptNoProxyKey,
+		config.AptMirrorKey,
+		config.SnapHTTPProxyKey, config.SnapHTTPSProxyKey,
+		config.SnapStoreAssertionsKey,
+		config.SnapStoreProxyKey, config.SnapStoreProxyURLKey,
+		config.CloudInitUserDataKey,
+		config.ContainerInheritPropertiesKey,
+	)
+	if err != nil {
+		return result, fmt.Errorf(
+			"cannot get model config keys when calculating container config: %w",
+			err,
+		)
+	}
+
+	enableOSRefreshUpdate, _ := strconv.ParseBool(modelConfig[config.EnableOSRefreshUpdateKey])
+	enableOSUpgrade, _ := strconv.ParseBool(modelConfig[config.EnableOSUpgradeKey])
+	result.UpdateBehavior = &params.UpdateBehavior{
+		EnableOSRefreshUpdate: enableOSRefreshUpdate,
+		EnableOSUpgrade:       enableOSUpgrade,
+	}
+	result.ProviderType = modelConfig[config.TypeKey]
+	result.SSLHostnameVerification, _ = strconv.ParseBool(modelConfig[config.SSLHostnameVerificationKey])
+	result.LegacyProxy = proxy.Settings{
+		Http:    modelConfig[config.HTTPProxyKey],
+		Https:   modelConfig[config.HTTPSProxyKey],
+		Ftp:     modelConfig[config.FTPProxyKey],
+		NoProxy: modelConfig[config.NoProxyKey],
+	}
+	result.JujuProxy = proxy.Settings{
+		Http:    modelConfig[config.JujuHTTPProxyKey],
+		Https:   modelConfig[config.JujuHTTPSProxyKey],
+		Ftp:     modelConfig[config.JujuFTPProxyKey],
+		NoProxy: modelConfig[config.JujuNoProxyKey],
+	}
+	result.AptProxy = proxy.Settings{
+		Http:    addSchemeIfMissing("http", getWithFallback(modelConfig, config.AptHTTPProxyKey, config.JujuHTTPProxyKey, config.HTTPProxyKey)),
+		Https:   addSchemeIfMissing("https", getWithFallback(modelConfig, config.AptHTTPSProxyKey, config.JujuHTTPSProxyKey, config.HTTPSProxyKey)),
+		Ftp:     addSchemeIfMissing("ftp", getWithFallback(modelConfig, config.AptFTPProxyKey, config.JujuFTPProxyKey, config.FTPProxyKey)),
+		NoProxy: aptNoProxy(modelConfig),
+	}
+	result.AptMirror = modelConfig[config.AptMirrorKey]
+	result.SnapProxy = proxy.Settings{
+		Http:  modelConfig[config.SnapHTTPProxyKey],
+		Https: modelConfig[config.SnapHTTPSProxyKey],
+	}
+	result.SnapStoreAssertions = modelConfig[config.SnapStoreAssertionsKey]
+	result.SnapStoreProxyID = modelConfig[config.SnapStoreProxyKey]
+	result.SnapStoreProxyURL = modelConfig[config.SnapStoreProxyURLKey]
+	result.CloudInitUserData, _ = ensureStringMaps(modelConfig[config.CloudInitUserDataKey])
+	result.ContainerInheritProperties = modelConfig[config.ContainerInheritPropertiesKey]
+	return result, nil
+}
+
+// addSchemeIfMissing adds a scheme to a URL if it is missing
+// Copied from github.com/juju/juju/environs/config
```

Comment:
I don't like the fact that we're dragging in all of these methods from environs/config, but what else can we do?